generic: platform/mikrotik: add NVMEM layout driver

robimarko opened the pull request:
Currently, information from MikroTik hard_config is only available via sysfs, meaning that we have to rely on userspace to for example setup MACs.

So, lets provide a basic NVMEM layout based driver to expose the same cells as sysfs driver exposes.

Do note that the we dont extract the WLAN caldata and BDF-s at this point.

robimarko committed:
generic: platform/mikrotik: add NVMEM layout driver

Currently, information from MikroTik hard_config is only available via
sysfs, meaning that we have to rely on userspace to for example setup MACs.

So, lets provide a basic NVMEM layout based driver to expose the same cells
as sysfs driver exposes.

Do note that the we dont extract the WLAN caldata and BDF-s at this point.

Reviewed-by: Thibaut VARÈNE <hacks@slashdirt.org>
Link: https://github.com/openwrt/openwrt/pull/15665
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@f00b4r0 Should be good for a start, makes setting MAC-s much easier and allows setting correct ones per DSA port.

john-tho commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+#include <linux/platform_device.h>
+
+#define RB_MAGIC_HARD		"Hard"
+#define RB_HARDCONFIG_LEN 	0x1000
```
This pending patch exports nvmem device size instead:
https://lore.kernel.org/lkml/20230718160652.26756-2-zajec5@gmail.com/

robimarko replied:
Well, the thing is that patch has not moved for almost a year and I dont want to carry it though it was my idea as well to fetch the info from the MTD device, however there is no way to get the underlying MTD device from a NVMEM layout device as they intentionally made the structure internal to force people to go through helpers.

f00b4r0 replied:
hard_config is 0x2000 on e.g. ipq4018 devices: this code will not work everywhere. Can't you get the size from DTB at a minimum? default to 0x1000 if not set?

robimarko replied:
We need to get the underlying MTD device here as only it has the size, there was no way to get its node as they walled of a the underlying NVMEM structure, but we can get the layout node and then get its parent (the partition node) and this works for getting the MTD size to make this dynamic.

john-tho replied:
I had not noticed `nvmem_dev_size` went into 6.8. and has been backported (thanks Rafał) into OpenWrt for some time: https://github.com/openwrt/openwrt/blob/2db6459ed4e27939e1b72472eb22a9b979599d39/target/linux/generic/backport-6.6/819-v6.8-0010-nvmem-core-add-nvmem_dev_size-helper.patch#L29 https://github.com/openwrt/openwrt/commit/04ed95d1f92058ec26826791e309ec3ae09ee625

so could use that to avoid doing the mtd size workarounds.

john-tho commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+	u8 magic[4];
+} __packed;
+
+struct hard_config_tlv {
+	u16 tag;
+	u16 len;
+} __packed;
```
I needed to swap these when big endian (ath79) (in my own very rough driver that does the same thing).
```
OFFSET      mikrotik/OEM/hapac2/mtd/hapac2_c454.img
--------------------------------------------------------------------------------
0x0008E000  48 61 72 64 1A 00 04 00 00 00 00 00 04 00 08 00 |Hard............|
0x0008E010  48 8F 5A
```
```
OFFSET      mikrotik/OEM/hapac/mtd/RouterBoot.img
--------------------------------------------------------------------------------
0x0000E000  64 72 61 48 00 04 00 1A 00 00 00 00 00 08 00 04 |draH............|
0x0000E010  48 8F 5A
```

robimarko replied:
Hm, I based this implementation on the existing platform drivers and it claimed that the hard_config content is cpu-endian.

I still need to dig out some MIPS board to check the situation.

f00b4r0 replied:
hard-config tags are CPU-endian, yes: https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/routerboot.c#L37

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
```
Quick nit, missing an SPDX-FileCopyrightText here (or a copyright mention in code comment), AIUI "Module Author" isn't enough.

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+	if (ret < 0)
+		return ret;
+
+	if (memcmp(hdr.magic, RB_MAGIC_HARD, sizeof(hdr.magic))) {
```
the magic value is CPU-endian, memcmp() will not work on both endiannesses since you're comparing byte by byte. You can simply do an u32 comparison, which is also much faster. See: https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/mtd/parsers/routerbootpart.c#L183 or https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c#L715-L716

robimarko replied:
Yeah, I moved to this as it avoids the need to care about endianess.

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+} __packed;
+
+/* ID values for hardware settings */
+#define RB_ID_FLASH_INFO		0x03
+#define RB_ID_MAC_ADDRESS_PACK		0x04
+#define RB_ID_BOARD_PRODUCT_CODE	0x05
+#define RB_ID_BIOS_VERSION		0x06
+#define RB_ID_SDRAM_TIMINGS		0x08
+#define RB_ID_DEVICE_TIMINGS		0x09
+#define RB_ID_SOFTWARE_ID		0x0A
+#define RB_ID_SERIAL_NUMBER		0x0B
+#define RB_ID_MEMORY_SIZE		0x0D
+#define RB_ID_MAC_ADDRESS_COUNT		0x0E
+#define RB_ID_HW_OPTIONS		0x15
+#define RB_ID_WLAN_DATA			0x16
+#define RB_ID_BOARD_IDENTIFIER		0x17
+#define RB_ID_PRODUCT_NAME		0x21
+#define RB_ID_DEFCONF			0x26
+#define RB_ID_BOARD_REVISION		0x27
```
I would move these defines to a common header that both this driver and the sysfs one can share.

robimarko replied:
Yeah, working on it now.

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,214 @@
+			}
+		}
+skip:
+		offset += sizeof(tlv) + tlv.len;
```
You seem to be partly reimplementing `routerboot_tag_find()` here and ignoring this bit:
https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c#L715-L716

while I understand that this is not a concern for your current use case, I wouldn't trust MikroTik not to break things behind your back: why not use `routerboot_tag_find()` directly and drastically simplify this code, especially as you're only interested in a single tag here?

robimarko replied:
I would prefer not to rely on any external code and keep NVMEM layout driver self-contained.

I added alignment for the offset to 32 bits(Hopefully correctly), thanks for the comment.

f00b4r0 replied:
Hmm, I'm not sure I would call `routerboot_tag_find()` "external code" in this context. I appreciate your desire to keep the code self-contained but duplicating code to achieve very similar functionality is generally not ideal from a maintenance and resource-usage PoV. This is your call though :)

f00b4r0 reviewed (changes requested):
Quick first pass, HTH

I can't seem to remember the reason why I gave up on NVMEM for routerboot: something couldn't be done there but I can't put my finger on it anymore. Maybe NVMEM has gotten to a point where this is no longer an issue :)

robimarko committed:
generic: platform/mikrotik: move hard config tag ID-s to a header

Move the hard config tag ID-s to a separate header so they can be reused
by the NVMEM driver as well.

Link: https://github.com/openwrt/openwrt/pull/15665
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
> I can't seem to remember the reason why I gave up on NVMEM for routerboot: something couldn't be done there but I can't put my finger on it anymore. Maybe NVMEM has gotten to a point where this is no longer an issue :)

I think the reason was wlan calibration, more precisely the issue that some hap ac2/3 boards have one file while others have 2 and they just changed it in the middle of production.

f00b4r0 commented:
> > I can't seem to remember the reason why I gave up on NVMEM for routerboot: something couldn't be done there but I can't put my finger on it anymore. Maybe NVMEM has gotten to a point where this is no longer an issue :)
> 
> I think the reason was wlan calibration, more precisely the issue that some hap ac2/3 boards have one file while others have 2 and they just changed it in the middle of production.

Ah yes, I think you're exactly right and that was the killer :)

robimarko commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,244 @@
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
```
No, NVMEM cells are not constantly in memory, the layout driver just finds them and populates the cell information, so then they are read directly from the underlying storage only when requested by code, so there is no memory being used otherwise.

john-tho commented on target/linux/generic/files/drivers/platform/mikrotik/Kconfig:
```diff
@@ -15,4 +15,10 @@ config MIKROTIK_RB_SYSFS
 	help
 	  This driver exposes RouterBoot configuration in sysfs.
 
+config NVMEM_LAYOUT_MIKROTIK
```
Could we please set or is unset this in the configs that use `CONFIG_MIKROTIK`. Interrupts kernel config in build for the new symbol currently.
```
git grep CONFIG_MIKROTIK=y
target/linux/ath79/mikrotik/config-default:CONFIG_MIKROTIK=y
target/linux/ipq40xx/mikrotik/config-default:CONFIG_MIKROTIK=y
target/linux/ramips/mt7621/config-6.6:CONFIG_MIKROTIK=y
```

robimarko replied:
Done

john-tho commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,244 @@
+ 	 * The nodes are cpu-endian on the flash media. The payload is cpu-endian when
+ 	 * applicable. Tag nodes are not ordered (by ID) on flash.
+ 	 */
+	while (offset < data_len) {
```
```suggestion
	while ((offset + sizeof(tlv)) <= data_len) {
```
Should only be a problem with bad configuration or bad data, but want to make sure the type length read we do is still within data.

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,244 @@
+ 	 * applicable. Tag nodes are not ordered (by ID) on flash.
+ 	 */
+	while (offset < data_len) {
+		memcpy(&tlv, data + offset, sizeof(tlv));
```
Another mem*() I missed: this is very slow. If you want to keep things structured you could use a union for the tlv and natively load a 32-bit value, this would be much faster. I might update routerboot_tag_find() to do just that in fact :)

f00b4r0 commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,244 @@
+			break;
+		}
+
+		if (tlv.tag != 0) {
```
Not sure about that: I expect like you that tag id 0 is invalid, but I have no proof of that. The tag chain actually ends with a null *word*, that is both tag and len being zero.
As a side note, your code is not checking for this end condition and may process random data as tags after the end marker, AFAICT.

robimarko commented:
@f00b4r0 Finally caught some time to rethink this and convert it to 32 direct access instead.

john-tho commented on target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c:
```diff
@@ -0,0 +1,219 @@
+			cell.bytes = ETH_ALEN;
+		else
+			cell.bytes = tlv_len;
+		cell.np = of_get_child_by_name(layout, cell.name);
```
Wants `of_node_put`: https://docs.kernel.org/devicetree/kernel-api.html#c.of_get_child_by_name
```suggestion
		cell.np = of_get_child_by_name(layout, cell.name);
		of_node_put(cell.np);
```

robimarko replied:
Hm, I am not sure that `of_node_put` should be used here as that node can/will be used later.
I know that other layout drivers do not use `of_node_put` on the cell node.

john-tho commented:
Oops, yes, sorry.
Of course nvmem will need a live reference to the cell…
-- 
John Thomson

f00b4r0 reviewed (approved):
Reviewed-by: Thibaut VARÈNE <hacks@slashdirt.org>

robimarko commented:
Thanks! Rebased on top of main and merged!
